x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java

```diff
@@ -208,47 +218,98 @@ static void getMemoryAndProcessors(
+        ) - totalAssignedProcessors(autoscalingContext);
         for (var modelAssignment : autoscalingContext.modelAssignments.entrySet()) {
             TrainedModelAssignment assignment = modelAssignment.getValue();
-            final int numberOfAllocations = assignment.getTaskParams().getNumberOfAllocations();
+            final int numberOfRequestedAllocations = assignment.getTaskParams().getNumberOfAllocations();
             final int numberOfThreadsPerAllocation = assignment.getTaskParams().getThreadsPerAllocation();
             final long estimatedMemoryUsage = assignment.getTaskParams().estimateMemoryUsageBytes();
+            final int numberOfTargetAllocationsOnExistingNodes = assignment.totalTargetAllocations();
+            final int numMissingAllocations = numberOfRequestedAllocations - numberOfTargetAllocationsOnExistingNodes;
+            final int numMissingProcessors = numMissingAllocations * numberOfThreadsPerAllocation;
+            int numExistingProcessorsToBeUsed = Math.min(numMissingProcessors, numberOfAvailableProcessors);
 
             if (AssignmentState.STARTING.equals(assignment.getAssignmentState()) && assignment.getNodeRoutingTable().isEmpty()) {
-
+                // this state occurs when the assignment has just been added and has not been assigned to any nodes
                 logger.debug(
                     () -> format(
                         "trained model [%s] lacks assignment , memory required [%d]",
                         modelAssignment.getKey(),
                         estimatedMemoryUsage
                     )
                 );
-
-                extraSingleNodeModelMemoryInBytes = Math.max(extraSingleNodeModelMemoryInBytes, estimatedMemoryUsage);
+                extraPerNodeModelMemoryBytes = Math.max(extraPerNodeModelMemoryBytes, estimatedMemoryUsage);
                 extraModelMemoryInBytes += estimatedMemoryUsage;
 
-                // if not low priority, check processor requirements
+                // if not low priority, check processor requirements.
                 if (Priority.LOW.equals(modelAssignment.getValue().getTaskParams().getPriority()) == false) {
-                    // as assignments can be placed on different nodes, we only need numberOfThreadsPerAllocation here
-                    extraSingleNodeProcessors = Math.max(extraSingleNodeProcessors, numberOfThreadsPerAllocation);
-                    extraProcessors += numberOfAllocations * numberOfThreadsPerAllocation;
+                    if (numMissingProcessors > numberOfAvailableProcessors) {
+                        // as assignments can be placed on different nodes, we only need numberOfThreadsPerAllocation here
+                        extraSingleNodeProcessors = Math.max(extraSingleNodeProcessors, numberOfThreadsPerAllocation);
+                        extraProcessors += numMissingProcessors - numExistingProcessorsToBeUsed;
```

Comment:
`numExistingProcessorsToBeUsed` can only be used once. If there are multiple model assignments those existing processors could have been taken by an earlier assignment.

Would initialising `extraProcessors` to `-numberOfAvailableProcessors ` and change this to `extraProcessors += numMissingProcessors;` work? You would need to check that `extraProcessors` is not negative later on. 

If you do that then I think the `if (numMissingProcessors > numberOfAvailableProcessors) {` can be deleted too.